src/rt-app.c

```diff
@@ -794,26 +803,44 @@ void *thread_body(void *arg)
 			log_timing(data->log_handler, curr_timing);
 
 		if (opts.ftrace)
-			log_ftrace(ft_data.marker_fd, "[%d] end loop %d phase %d step %d",
-				   data->ind, i, j, loop);
-
-		loop++;
-		if (loop == pdata->loop) {
-			loop = 0;
-
-			j++;
-			if (j == data->nphases) {
-				j = 0;
-				i++;
+			log_ftrace(ft_data.marker_fd, "[%d] end loop %d phase %d",
+				   data->ind, phase_loop, phase);
+
+		phase_loop++;
+		/* Reached the specified number of loops for this phase. */
+		if (phase_loop == pdata->loop) {
+			/* Check if we have a following phase to execute */
+			phase++;
+			if (phase != data->nphases) {
+				/*
+				 * Yes, we have it, so execute it
+				 * (and reset loop count for it)
+				 */
+				phase_loop = 0;
+				pdata = &data->phases[phase];
+			} else if (data->loop == -1 || thread_loop < data->loop - 1) {
+				/*
+				 * Phases are finished, but we might
+				 *
+				 * - have to execute indefinetly
+				 *   (data->loop, i.e. task level "loop", == -1)
+				 * - perform a number of task level iterations of
+				 *   the same phase(s)
+				 *
+				 * In both cases, start over from first phase.
+				 *
+				 */
+				phase_loop = 0;
+				phase = 0;
+				thread_loop++;
+				pdata = &data->phases[0];
```

Comment:
I think we can only do something like

```
diff --git a/src/rt-app.c b/src/rt-app.c
index a7f26c9..1a46c83 100644
--- a/src/rt-app.c
+++ b/src/rt-app.c
@@ -817,7 +821,6 @@ void *thread_body(void *arg)
 				 * (and reset loop count for it)
 				 */
 				phase_loop = 0;
-				pdata = &data->phases[phase];
 			} else if (data->loop == -1 || thread_loop < data->loop - 1) {
 				/*
 				 * Phases are finished, but we might
@@ -833,8 +836,14 @@ void *thread_body(void *arg)
 				phase_loop = 0;
 				phase = 0;
 				thread_loop++;
-				pdata = &data->phases[0];
 			}
+
+			/*
+			 * If we reset phase_loop it means that phase now
+			 * points to a valid phase (next or first).
+			 */
+			if (!phase_loop)
+				pdata = &data->phases[phase];
 		}
 
 		log_idx++;
```
Otherwise we might point pdata to unknown memory after last iteration and phase_loop != pdata->loop check might segfault.